ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_15/EventQueryServiceIT.scala

```diff
@@ -0,0 +1,269 @@
+          contractKey = Some(key),
+          templateId = Some(TextKey.id.unwrap),
+          requestingParties = Tag.unsubst(immutable.Seq(party)),
+        )
+      )
+    } yield {
+      val actual = assertDefined(events.archiveEvent, "Expected a archived event")
+      assertEquals("Looked up event should match the transaction event", actual, expected)
+    }
+  })
+
+  test(
+    "TXEventsByContractKeyNoKey",
+    "No events are returned for a non existent contract key",
+    allocate(SingleParty),
+  )(implicit ec => { case Participants(Participant(ledger, party)) =>
+    val key = makeTextKeyKey(party, "non existent key")
+
+    for {
+      events <- ledger.getEventsByContractKey(
+        GetEventsByContractKeyRequest(
+          contractKey = Some(key),
+          templateId = Some(TextKey.id.unwrap),
+          requestingParties = Tag.unsubst(immutable.Seq(party)),
+        )
+      )
+    } yield {
+      assertIsEmpty(Seq(events.createEvent, events.archiveEvent).flatten[GeneratedMessage])
+    }
+  })
+
+  test(
+    "TXEventsByContractKeyNotVisible",
+    "No events are returned for a non visible contract key",
+    allocate(TwoParties),
+  )(implicit ec => { case Participants(Participant(ledger, party, notTheSubmittingParty)) =>
+    val nonVisibleKey = "non visible key"
+    val key = makeTextKeyKey(party, nonVisibleKey)
+
+    for {
+      _ <- ledger.submitAndWaitForTransaction(
+        ledger.submitAndWaitRequest(party, TextKey(party, nonVisibleKey, Nil).create.command)
+      )
+
+      events <- ledger.getEventsByContractKey(
+        GetEventsByContractKeyRequest(
+          contractKey = Some(key),
+          templateId = Some(TextKey.id.unwrap),
+          requestingParties = immutable.Seq(notTheSubmittingParty.unwrap),
+        )
+      )
+    } yield {
+      assertIsEmpty(Seq(events.createEvent, events.archiveEvent).flatten[GeneratedMessage])
+    }
+  })
+
+  test(
+    "TXEventsByContractKeyEndExclusive",
+    "Should return event prior to the end exclusive event",
+    allocate(SingleParty),
+  )(implicit ec => { case Participants(Participant(ledger, party)) =>
+    val exercisedKey = "paging key"
+    val key = makeTextKeyKey(party, exercisedKey)
+
+    def getNextResult(endExclusiveEventId: String): Future[Option[String]] = {
+      ledger
+        .getEventsByContractKey(
+          GetEventsByContractKeyRequest(
+            contractKey = Some(key),
+            templateId = Some(TextKey.id.unwrap),
+            requestingParties = Tag.unsubst(immutable.Seq(party)),
+            endExclusiveEventId = endExclusiveEventId,
+          )
+        )
+        .map(_.createEvent.map(_.eventId))
+    }
+
+    for {
+      textKeyCid1 <- ledger.create(party, TextKey(party, exercisedKey, Nil))
+      _ <- ledger.submitAndWaitForTransaction(
+        ledger.submitAndWaitRequest(party, textKeyCid1.exerciseTextKeyChoice().command)
+      )
+      textKeyCid2 <- ledger.create(party, TextKey(party, exercisedKey, Nil))
```

Comment:
A more common pattern is when a consuming choice creates a new instance of a contract with a key. It would be interesting to assert that the back-tracking works as desired for both `preconsuming` and `postconsuming` flavors.